Add UnicodeFuncs with safe-to-implement traits

sffc opened the pull request:
This PR makes it easier for clients to implement and use `hb_unicode_funcs_t`.

We want to implement these in ICU4X: https://github.com/unicode-org/icu4x/issues/2514

CC @hsivonen @echeran

sffc commented on harfbuzz/src/unicode_funcs.rs:
````diff
@@ -0,0 +1,357 @@
+/// b.guess_segment_properties();
+/// assert_eq!(b.get_direction(), Direction::RTL);
+/// assert_eq!(b.get_script(), sys::HB_SCRIPT_ARABIC);
+/// ```
````
Note: Currently, running this test results in the following error

```
stderr:
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `LTR`,
 right: `RTL`', src/unicode_funcs.rs:20:1
stack backtrace:
   0:     0x559bce29ab5d - std::backtrace_rs::backtrace::libunwind::trace::h22893a5306c091b4
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x559bce29ab5d - std::backtrace_rs::backtrace::trace_unsynchronized::h29c3bc6f9e91819d
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x559bce29ab5d - std::sys_common::backtrace::_print_fmt::he497d8a0ec903793
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/sys_common/backtrace.rs:66:5
   3:     0x559bce29ab5d - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h9c2a9d2774d81873
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/sys_common/backtrace.rs:45:22
   4:     0x559bce2b482c - core::fmt::write::hba4337c43d992f49
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/fmt/mod.rs:1194:17
   5:     0x559bce298eb1 - std::io::Write::write_fmt::heb73de6e02cfabed
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/io/mod.rs:1655:15
   6:     0x559bce29c305 - std::sys_common::backtrace::_print::h63c8b24acdd8e8ce
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/sys_common/backtrace.rs:48:5
   7:     0x559bce29c305 - std::sys_common::backtrace::print::h426700d6240cdcc2
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/sys_common/backtrace.rs:35:9
   8:     0x559bce29c305 - std::panicking::default_hook::{{closure}}::hc9a76eed0b18f82b
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:295:22
   9:     0x559bce29bfb9 - std::panicking::default_hook::h2e88d02087fae196
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:314:9
  10:     0x559bce29c852 - std::panicking::rust_panic_with_hook::habfdcc2e90f9fd4c
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:698:17
  11:     0x559bce29c737 - std::panicking::begin_panic_handler::{{closure}}::he054b2a83a51d2cd
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:588:13
  12:     0x559bce29b014 - std::sys_common::backtrace::__rust_end_short_backtrace::ha48b94ab49b30915
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/sys_common/backtrace.rs:138:18
  13:     0x559bce29c469 - rust_begin_unwind
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:584:5
  14:     0x559bce27f383 - core::panicking::panic_fmt::h366d3a309ae17c94
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panicking.rs:143:14
  15:     0x559bce2b3b18 - core::panicking::assert_failed_inner::h92c01a43e8afd9c2
  16:     0x559bce27fc86 - core::panicking::assert_failed::h509e1b6678eef5a0
  17:     0x559bce28146b - rust_out::main::_doctest_main_src_unicode_funcs_rs_291_0::he894c01431e2c97b
  18:     0x559bce281166 - rust_out::main::hfc3225cdebcc9762
  19:     0x559bce27f803 - core::ops::function::FnOnce::call_once::h6e5409427dfdbb1d
  20:     0x559bce27f649 - std::sys_common::backtrace::__rust_begin_short_backtrace::h97d1943f78598fa4
  21:     0x559bce27f6b9 - std::rt::lang_start::{{closure}}::hd9e98fd00335415b
  22:     0x559bce296c3e - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::had4f69b3aefb47a8
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/ops/function.rs:259:13
  23:     0x559bce296c3e - std::panicking::try::do_call::hf2ad5355fcafe775
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:492:40
  24:     0x559bce296c3e - std::panicking::try::h0a63ac363423e61e
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:456:19
  25:     0x559bce296c3e - std::panic::catch_unwind::h18088edcecb8693a
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panic.rs:137:14
  26:     0x559bce296c3e - std::rt::lang_start_internal::{{closure}}::ha7dad166dc711761
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/rt.rs:128:48
  27:     0x559bce296c3e - std::panicking::try::do_call::hda0c61bf3a57d6e6
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:492:40
  28:     0x559bce296c3e - std::panicking::try::hbc940e68560040a9
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:456:19
  29:     0x559bce296c3e - std::panic::catch_unwind::haed0df2aeb3fa368
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panic.rs:137:14
  30:     0x559bce296c3e - std::rt::lang_start_internal::h9c06694362b5b80c
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/rt.rs:128:20
  31:     0x559bce27f6a1 - std::rt::lang_start::h468e868f8b4aa277
  32:     0x559bce281613 - main
  33:     0x7fefcf16718a - __libc_start_call_main
                               at ./csu/../sysdeps/nptl/libc_start_call_main.h:58:16
  34:     0x7fefcf167245 - __libc_start_main_impl
                               at ./csu/../csu/libc-start.c:381:3
  35:     0x559bce27f571 - _start
  36:                0x0 - <unknown>


```

sffc commented on harfbuzz/src/unicode_funcs.rs:
```diff
@@ -0,0 +1,352 @@
+pub trait GeneralCategoryFunc {
+    /// Given a code point, return the general category as a
+    /// [`hb_unicode_general_category_t`].
+    fn general_category(&self, ch: u32) -> hb_unicode_general_category_t;
```
Thought: Perhaps these return values should not use `harfbuzz-sys` types such that these traits can be implemented without a `harfbuzz-sys` dependency (see https://github.com/unicode-org/icu4x/issues/3175)

sffc commented:
:wave: Can I get a signal on whether this is the right direction and the shape of what you would approve?

Manishearth commented:
cc @jdm I'm happy to review this but there's a bit of a conflict of interest, would rather get a :+1: from other Servo people on whether this is a direction we want to go in.

Perhaps we should mention this at the next TSC?


(ICU4X may also be happy to take on some of the burden of maintaining this crate, the way we are doing with unicode-bidi)

mrobinson commented:
This seems fine to me, though I wonder if it is ever the case that someone would implement only a subset of the functions? It would be simpler to only expose a single trait for implementing functions if it is always the case that you implement all of them.

Manishearth commented:
One thing it helps is that it allows people to separate different data sources into different crates and implementors, even allowing for mixing. 

Unclear if that level of granularity is desired, but it's a potential source of flexibility. (@sffc, thoughts?)

sffc commented:
> This seems fine to me, though I wonder if it is ever the case that someone would implement only a subset of the functions? It would be simpler to only expose a single trait for implementing functions if it is always the case that you implement all of them.

There could be use cases. For example, maybe you want to overwrite your Script data but leave the Normalizer alone. Harfbuzz seems to make no requirement that all six functions be implemented.

This is ready for re-review.

Manishearth commented:
r? @hsivonen

Ideally someone with familiarity with harfbuzz reviews this, I can get it merged after that.

waywardmonkeys commented:
@sffc If you rebase forward, the underlying macOS build issue in CI is fixed.

hsivonen commented:
I have a couple of concerns:

Concern 1:

Does this add function dereferences for each call compared to the case without this abstraction? I _think_ it does when comparing against a case where a callback without this abstraction wouldn't use the user data pointer, but I _think_ it does not when comparing against a case where the callback without this abstraction would use the user data pointer. Is that correct? (I think it's OK if this abstraction doesn't add pointer chasing compared to the case where the user data pointer is used anyhow.)

Concern 2:

It's a bit odd for the traits to be simultaneously high-level enough to be implementable without `unsafe` and low-level and/or HarfBuzz-idiosyncratic in the argument and return types.

Unicode Scalar Values are represented as `u32` instead of `char`. As noted in https://github.com/unicode-org/icu4x/issues/2840 , if we took the position that calling `hb_buffer_add_codepoints` is conceptually like calling Rust's `core::str::from_utf8_unchecked`, it would be proper for the trait method signatures to use `char` for Unicode Scalar Value and for the glue code provided here to perform an unchecked conversion from `hb_codepoint_t` to `char` when receiving `hb_codepoint_t` from HarfBuzz.

Canonical Combining Class has numeric values defined by Unicode and the [Unicode stability policy](https://www.unicode.org/policies/stability_policy.html) guarantees their possible range to fit in the `u8` range. (Specifically, the guaranteed value space is one item smaller than the value space of `u8`, since 255 is guaranteed not to be used as a canonical combining class.) As `CombiningClassFunc` involves passing Canonical Combining Class only in the direction of Rust to HarfBuzz, it would make sense to use `u8` in the trait and have the glue code zero-extend it in order to avoid C-dependent types in the trait signatures when possible.

The Unicode stability policy guarantees that the number of items for General Category won't change and HarfBuzz has assigned all the numeric values to fit in `u8`. Unfortunately, the numeric values that HarfBuzz expects are not defined by Unicode but by HarfBuzz, so leaving the type as is would require the trait implementor to get the HarfBuzz numeric assignments from somewhere. `icu_harfbuzz` currently gets them by depending on `harfbuzz-sys`. To the extent the point of this PR is to allow the trait implementor not to directly depend on `harfbuzz-sys`, that's not ideal.

Since the HarfBuzz numeric assignments cannot change, I suggest creating an enum in this crate alongside the traits with such that the enum representation is `u8` with explicitly-assigned HarfBuzz-compatible values and having the glue code zero-extend the `enum` into `c_uint`. By making the Rust-visible `enum` representation have the size of `u8` would enable the [permutation table in `icu_harfbuzz`](https://github.com/unicode-org/icu4x/blob/83550ce5689beeffc38adccfe34c1f96c195856c/experimental/harfbuzz/src/lib.rs#L83) to become type-safe without becoming larger.

sffc commented:
> I have a couple of concerns:
> 
> Concern 1:
> 
> Does this add function dereferences for each call compared to the case without this abstraction? I _think_ it does when comparing against a case where a callback without this abstraction wouldn't use the user data pointer, but I _think_ it does not when comparing against a case where the callback without this abstraction would use the user data pointer. Is that correct? (I think it's OK if this abstraction doesn't add pointer chasing compared to the case where the user data pointer is used anyhow.)

Everything is super generic so I hope the compiler will inline these abstractions. For example, in this code:

```rust
        let compose_ptr: *mut F = Box::into_raw(f);
        extern "C" fn impl_compose<F: ComposeFunc>(
            // ...
            user_data: *mut c_void,
        ) -> hb_bool_t {
            let result = unsafe { &*(user_data as *mut F) }
                .compose(hb_codepoint_t_to_char(a), hb_codepoint_t_to_char(b));
            // ...
        }
 ```

the compiler should be able to figure out the address of `.compose` relative to the address of `user_data`, I think.


> Concern 2:
> 
> It's a bit odd for the traits to be simultaneously high-level enough to be implementable without `unsafe` and low-level and/or HarfBuzz-idiosyncratic in the argument and return types.
> 
> Unicode Scalar Values are represented as `u32` instead of `char`. As noted in [unicode-org/icu4x#2840](https://github.com/unicode-org/icu4x/issues/2840) , if we took the position that calling `hb_buffer_add_codepoints` is conceptually like calling Rust's `core::str::from_utf8_unchecked`, it would be proper for the trait method signatures to use `char` for Unicode Scalar Value and for the glue code provided here to perform an unchecked conversion from `hb_codepoint_t` to `char` when receiving `hb_codepoint_t` from HarfBuzz.

Changed from `u32` to `char`: 5b8a4aaafcb8b18bc590443545dac2ac393e7643

> Canonical Combining Class has numeric values defined by Unicode and the [Unicode stability policy](https://www.unicode.org/policies/stability_policy.html) guarantees their possible range to fit in the `u8` range. (Specifically, the guaranteed value space is one item smaller than the value space of `u8`, since 255 is guaranteed not to be used as a canonical combining class.) As `CombiningClassFunc` involves passing Canonical Combining Class only in the direction of Rust to HarfBuzz, it would make sense to use `u8` in the trait and have the glue code zero-extend it in order to avoid C-dependent types in the trait signatures when possible.

Changed from `c_uint` to `u8`: 39af9196351f3eaff25d93d34ff8dd6466db56be

> The Unicode stability policy guarantees that the number of items for General Category won't change and HarfBuzz has assigned all the numeric values to fit in `u8`. Unfortunately, the numeric values that HarfBuzz expects are not defined by Unicode but by HarfBuzz, so leaving the type as is would require the trait implementor to get the HarfBuzz numeric assignments from somewhere. `icu_harfbuzz` currently gets them by depending on `harfbuzz-sys`. To the extent the point of this PR is to allow the trait implementor not to directly depend on `harfbuzz-sys`, that's not ideal.
> 
> Since the HarfBuzz numeric assignments cannot change, I suggest creating an enum in this crate alongside the traits with such that the enum representation is `u8` with explicitly-assigned HarfBuzz-compatible values and having the glue code zero-extend the `enum` into `c_uint`. By making the Rust-visible `enum` representation have the size of `u8` would enable the [permutation table in `icu_harfbuzz`](https://github.com/unicode-org/icu4x/blob/83550ce5689beeffc38adccfe34c1f96c195856c/experimental/harfbuzz/src/lib.rs#L83) to become type-safe without becoming larger.

Changed from `c_uint` to a `#[repr(u8)]` enum with Harfbuzz's values: 6868c9751e22706a78c0d0a1f4f3cad701649f59

sffc commented:
@waywardmonkeys can you kick CI?

hsivonen commented on harfbuzz/src/traits.rs:
```diff
@@ -0,0 +1,89 @@
+/// [`hb_unicode_general_category_t`]: crate::sys::hb_unicode_general_category_t
+#[repr(u8)]
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+#[allow(non_camel_case_types)] // the names are defined by Unicode
```
I suggest dropping the underscores like ICU4X does. That way, the names would be idiomatic for a Rust enum. Even if the names are defined by Unicode, it's pretty obvious to the reader what the applied name transformation would be if the names were idiomatic to Rust.

hsivonen replied:
Also, note the casing of the letter "u" in `PrivateUse`/`Private_Use`.

hsivonen reviewed (changes requested):
Thanks. Looking very good, though I have one nit about using Rust-idiomatic naming for `GeneralCategory` items.

waywardmonkeys commented:
> @waywardmonkeys can you kick CI?

I can't ... I didn't get commit privs back after losing them.

waywardmonkeys commented:
I don't understand why this `harfbuzz` would ever be built without `harfbuzz-sys` (at least in such a way that all of that code is removed). That makes it seem like some of this code should be in a third crate and not part of `harfbuzz` or `harfbuzz-sys`.

Similarly, I'm not sure what the use case is for `no-std`, but if that's what we want / need for `harfbuzz` / `harfbuzz-sys`, then I think we should do that as a separate commit with a clear history and rationale.

sffc commented:
> I don't understand why this `harfbuzz` would ever be built without `harfbuzz-sys` (at least in such a way that all of that code is removed). That makes it seem like some of this code should be in a third crate and not part of `harfbuzz` or `harfbuzz-sys`.

My mental model is that `harfbuzz` contains Rust abstractions over `harfbuzz-sys`. It is in the realm of responsibility of `harfbuzz` to define traits when the underlying C library has a trait-like abstraction, which is what we find in UnicodeFuncs.

The traits do not themselves invoke any C functions, which is why this PR makes `harfbuzz-sys` an optional dependency of `harfbuzz`.

> Similarly, I'm not sure what the use case is for `no-std`, but if that's what we want / need for `harfbuzz` / `harfbuzz-sys`, then I think we should do that as a separate commit with a clear history and rationale.

Crates that don't always need `std` should have an `std` feature to be more compatible with the `no_std` ecosystem. ICU4X is an example client that needs `no_std`.

waywardmonkeys commented:
> The traits do not themselves invoke any C functions, which is why this PR makes `harfbuzz-sys` an optional dependency of `harfbuzz`.

Why would someone use `harfbuzz` without using Harfbuzz?

The point is that it defines things like blobs and buffers and all of that (and should define so much more, like `harfbuzz_rs` does).

I wouldn't approve this without understanding that. (Although despite me writing most of the commits here over the last few years, I don't know what say I actually have.)

> Crates that don't always need `std` should have an `std` feature to be more compatible with the `no_std` ecosystem. ICU4X is an example client that needs `no_std`.

That seems reasonable. But it should be a separate PR that handles it completely.

I'll submit a PR that updates the generated bindings to use `core` since I have an environment for building those readily available, but the `harfbuzz` side should be done as a separate PR.

sffc commented:
> > The traits do not themselves invoke any C functions, which is why this PR makes `harfbuzz-sys` an optional dependency of `harfbuzz`.
> 
> Why would someone use `harfbuzz` without using Harfbuzz?
> 
> The point is that it defines things like blobs and buffers and all of that (and should define so much more, like `harfbuzz_rs` does).
> 
> I wouldn't approve this without understanding that. (Although despite me writing most of the commits here over the last few years, I don't know what say I actually have.)

It's so that a crate implementing these traits, like ICU4X, does not need to include any `sys` crates in its dependency tree.

For example, running `cargo test --all-features` in ICU4X currently requires having `cmake` available for building harfbuzz, which is a pain point (https://github.com/unicode-org/icu4x/issues/3175).

waywardmonkeys commented:
The need for `cmake` is due to `freetype` which is an optional dependency. You should be able to build this with only `pkg-config` and `cc` and we could restructure things slightly to do away with `cc` (it isn't used when `pkg-config` is) as a required dependency and make it only required when `bundled` is enabled.

But I still think that perhaps having a `harfbuzz-unicode-funcs` crate that provides the bare minimum there and then doing a re-export of it from `harfbuzz` would be cleaner and less maintenance going forward (as `harfbuzz` needs to grow a lot of other features to be useful and it seems wrong to predicate them all on `harfbuzz-sys` being enabled).

sffc commented:
Okay, how about naming it `harfbuzz-traits`? I can make these changes once #232 lands.

waywardmonkeys commented:
That sounds good. And now that I understand better, that is a good place perhaps for the draw traits as well when we support that.

sffc commented:
I merged this PR to sit on top of https://github.com/servo/rust-harfbuzz/pull/232 so it is ready once #232 lands.

hsivonen commented on harfbuzz-traits/src/unicode_funcs_traits.rs:
```diff
@@ -0,0 +1,89 @@
+    Control = 0,
+    Format = 1,
+    Unassigned = 2,
+    Private_use = 3,
```
`use` should be `Use` with upper-case `U`.

I still think the underscores in these values should be removed the way ICU4X removes the upstream Unicode underscores to get idiomatic Rust enum values.

waywardmonkeys replied:
I agree with both of those requests, fwiw.

hsivonen reviewed (changes requested):
Re-raising the previous nits. I trust the substance of the rest of the code hasn't changed--at least that's how it appeared from a cursory look.

waywardmonkeys reviewed (changes requested):
I think that `harfbuzz` should re-export the traits crate, like it does for `harfbuzz-sys`, so that people who do depend on `harfbuzz` can get access to the same version that their `harfbuzz` dependency is using.

I might be wrong, so say so if I am!

waywardmonkeys commented:
I think this has resolved all of my comments and concerns! If no one has other issues, I'll merge in about 12 hours.
